### DISCUSSION TITLE
Expose a react-router mechanism for use in catch-all routes

### DISCUSSION THREAD

**Discussion by Francesco-Lanciana** [vercel/next.js#23844] [Ideas]
Hello!

## The problem with file based routing
I have found that file-based really isn't the right fit when it comes to dealing with complex pages where **state needs go be maintained across multiple sub-routes** and/or **each sub-route needs the same data from the DB**. File based routing makes it really hard to maintain state across routes and it forces you to hit the API on each page even if the previous page had already loaded all the data you needed. The only options are to: 
1. Have some shared global state that sits in `_app.jsx` but this can start to get really messy as your app scales
2. Have some kind of caching mechanism but this isn't suited for cases when you want to maintain state across specific routes that shouldn't be persisted until a user action is performed (e.g. editing the same object over multiple routes that is persisted when the user clicks save).

## What would fix it
Ideally we should be able to switch to client side routing using something similar to react-router when we have a complex page with multiple sub-routes that need to share state. While NextJs has introduced the concept of catch-all route pages (`[[...page]].jsx`) it hasn't added any react-router like mechanism for defining sub-routes in this catch-all component. 

## My current workaround
Without a solution like react-router we are left to build our own routing mechanisms in catch-all routes. This gets really annoying when the catch-all route sits inside a dynamic route, e.g. `[workspace]/cards/[[...card]].jsx`. I'm currently using regex to  return certain components based off the current route.
```
if (/(.+)\/cards\/view/.test(router.asPath) return <CardViewSection />
else if ...
```

## Example of an app where this would come in handy
I'm currently building an app that has a page that can be in several different modes, where each mode should have it's own URL and there is a need to share state between routes. 
![image](https://user-images.githubusercontent.com/20393195/112058602-e48dca00-8bae-11eb-8790-378ca093ade9.png)

When in edit mode a user should be able to navigate between the tabs (Card View, Social Preview Links, QR Code, and Mobile Contact Preview), where each tab will allow them to edit a different part of the card. A user can then save these changes by clicking save.

----
Would be great to see an officially endorsed solution to this problem @Timer @timneutkens. Just seems like a use case that would be easy to support but really isn't mentioned in the docs. 

### Related resources/links
[Building a SPA with NextJs](https://colinhacks.com/essays/building-a-spa-with-nextjs)
[Issue: Use NextJs with React Router 4](https://github.com/vercel/next.js/issues/1632)
[Persistent Layout patterns in NextJs](https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/)

**Comment by tappleby** [comment]
@Francesco-Lanciana Did you end up finding any good solutions for this? I have found myself wanting a similar feature.

**Reply by Francesco-Lanciana** [reply]
@tappleby unfortunately there isn't really a good solution for this that I know of. I ended up essentially making a very lightweight custom solution using the matchit library. Here's a stripped down example of what I ended up with that let me handle dynamic routes.

```javascript
import type { NextPage } from "next";
import { exec, match, parse } from "matchit";
...
type RouteInfo =
    | { pageMode: "edit"; params: { id: string } }
    | { pageMode: "view"; params: { id: string } }
    | { pageMode: "create"; params: { id: undefined } }
    | { pageMode: "invalid"; params: { id: undefined } };

function getPathMatch(
    pathname: string,
    routes: string[],
    onMatch: (matchingRoute: string | undefined, values: Record<string, string>) => RouteInfo
): RouteInfo {
    const parsedRoutes = routes.map(parse);
    const matches = match(pathname, parsedRoutes);
    const values = exec(pathname, matches);

    return onMatch(matches[0].old, values);
}

const ACCEPTED_ROUTES = {
    NEW: ":workspaceId/presets/new",
    EDIT: ":workspaceId/presets/:id/edit",
    PRESET: ":workspaceId/presets/:id",
};

function useRouteInfo(): CardPresetPageRouteInfo {
    const router = useRouter();

    const routeInfo = useMemo(
        () =>
            getPathMatch(
                router.asPath.split("?")[0], // Remove query params
                Object.values(ACCEPTED_ROUTES),
                (matchingRoute, values) => {
                    switch (matchingRoute) {
                        case ACCEPTED_ROUTES.PRESET:
                            return {
                                pageMode: "view",
                                params: { id: values.id },
                            };
                       case ACCEPTED_ROUTES.EDIT:
                            return {
                                pageMode: "edit",
                                params: { id: values.id },
                            };
                        case ACCEPTED_ROUTES.NEW: {
                            return {
                                pageMode: "create",
                                params: { id: undefined },
                            };
                        }
                        default:
                            return { pageMode: "invalid", params: { id: undefined } };
                    }
                }
            ),
        [router.asPath, router.query.from, router.query.mode, router.query.tab]
    );

    return routeInfo;
}

...

const ExamplePage: NextPage = () => {
    const routeInfo = useRouteInfo();

    return (
        {routeInfo.pageMode === "view" && (
            <ViewSection />
        )}
        {routeInfo.pageMode === "edit" && (
            <EditSection />
        )}
        {routeInfo.pageMode === "create" && (
            <CreateSection />
        )}
    )
}
```

Essentially I created a standardised routeInfo object and based off that I would decide which section/version of the page I wanted to render, where each version of the page was just a normal React component. 

@Timer @timneutkens I would still love a standardised react-router type solution, is there anything planned for this? Right now I think this and and Incremental Static Regeneration that can be triggered via an API are two of the biggest things NextJs is missing in order to be a solution I can use for just about everything.
